Flaky panic in concurrent AppendToStream error handling

I have seen this panic a couple of times in our integration tests:

```
{"level":"error","msg":"subscription has dropped. Reason: [rpc error: code = Unavailable desc = error reading from server: EOF]"}
panic: send on closed channel

goroutine 6833 [running]:
github.com/EventStore/EventStore-Client-Go/v4/esdb.(*grpcClient).handleError(0xc0007bc120, 0xc0003a1830, 0xec7f80?, 0xc000649b90, {0x101a620, 0xc00006ecf8})
	/home/runner/go/pkg/mod/github.com/!event!store/!event!store-!client-!go/v4@v4.1.0/esdb/impl.go:78 +0x5f8
github.com/EventStore/EventStore-Client-Go/v4/esdb.(*Client).AppendToStream(0xc000891e70, {0x1024240, 0x1691160}, {0xc000136f80, 0x69}, {{0x1018180, 0x1691160}, 0x0, 0x0, 0x0}, ...)
	/home/runner/go/pkg/mod/github.com/!event!store/!event!store-!client-!go/v4@v4.1.0/esdb/client.go:87 +0xbae
project-othermodule/eventstore.ESDBCoreService.AppendToStream(...)
	/home/runner/work/someproject/someproject/backend/go-common/eventstore/core.request.go:331
project-othermodule/eventstore/service.ESDBDataService.updateDataProgress.func1()
	/home/runner/work/someproject/someproject/backend/go-common/eventstore/service/service.progress.go:170 +0x4d7
created by project-othermodule/eventstore/service.ESDBDataService.updateDataProgress in goroutine 6793
	/home/runner/work/someproject/someproject/backend/go-common/eventstore/service/service.progress.go:147 +0x234
FAIL eventstore/service.TestFlakyGraphqlReads (-1.00s)
=== RUN   TestFlakyGraphqlReads/DoesNotGetErrorStreamNotFound
```

The line that panics from `impl.go` in `handleError()`:
<img width="399" alt="Pasted Graphic" src="https://github.com/user-attachments/assets/160ab5ba-4a1e-45b0-8d5d-59d3559b0325">

Just before the panic we have a failing `AppendToStream`, that returns an error, that i'm afraid i have not logged. I suspect there is a problem with intermittent network errors in the github runners, because we get quite a few random flaky tests at the moment (we're running integration tests against eventstore via testcontainers-go).

Any way i'm not completely sure on the order of calls, but the net result is that `func (client *grpcClient) close() {}` must have been called _before_ `func (client *grpcClient) handleError()` which causes the panic above.

Some details about the test i'm running:
- I'm testing a complex service doing both reads, appends and subscribes
- I'm using multiple esdb clients from 5 different go routines (due to https://discuss.eventstore.com/t/grpc-connection-pool-for-subscriptions/5210)
- I'm _not_ explicitly closing any of those clients
- I might cancel contexts to the different esdb operations

I see there has already been some mitigation of concurrency errors with the `closeFlag` in the `type grpcClient struct`. I don't think it's sufficient though. If `AppendToStream` are to be thread safe i don't really see any way around a regular `sync.Mutex` being locked on all writes to the `channel` of the `grpcClient struct` as well as on closure.

Let me know if you need any more details, thanks in advance

--- a/esdb/client.go
+++ b/esdb/client.go
@@ -62,14 +62,14 @@ func (client *Client) AppendToStream(
 
 	appendOperation, err := streamsClient.Append(ctx, callOptions...)
 	if err != nil {
-		err = client.grpcClient.handleError(handle, headers, trailers, err)
+		err = client.grpcClient.handleError(handle, trailers, err)
 		return nil, fmt.Errorf("could not construct append operation. Reason: %w", err)
 	}
 
 	header := toAppendHeader(streamID, opts.ExpectedRevision)
 
 	if err := appendOperation.Send(header); err != nil {
-		err = client.grpcClient.handleError(handle, headers, trailers, err)
+		err = client.grpcClient.handleError(handle, trailers, err)
 		return nil, fmt.Errorf("could not send append request header. Reason: %w", err)
 	}
 
@@ -81,14 +81,14 @@ func (client *Client) AppendToStream(
 		}
 
 		if err = appendOperation.Send(appendRequest); err != nil {
-			err = client.grpcClient.handleError(handle, headers, trailers, err)
+			err = client.grpcClient.handleError(handle, trailers, err)
 			return nil, fmt.Errorf("could not send append request. Reason: %w", err)
 		}
 	}
 
 	response, err := appendOperation.CloseAndRecv()
 	if err != nil {
-		return nil, client.grpcClient.handleError(handle, headers, trailers, err)
+		return nil, client.grpcClient.handleError(handle, trailers, err)
 	}
 
 	result := response.GetResult()
@@ -249,7 +249,7 @@ func (client *Client) DeleteStream(
 	deleteRequest := toDeleteRequest(streamID, opts.ExpectedRevision)
 	deleteResponse, err := streamsClient.Delete(ctx, deleteRequest, callOptions...)
 	if err != nil {
-		err = client.grpcClient.handleError(handle, headers, trailers, err)
+		err = client.grpcClient.handleError(handle, trailers, err)
 		return nil, fmt.Errorf("failed to perform delete, details: %w", err)
 	}
 
@@ -280,7 +280,7 @@ func (client *Client) TombstoneStream(
 	tombstoneResponse, err := streamsClient.Tombstone(ctx, tombstoneRequest, callOptions...)
 
 	if err != nil {
-		err = client.grpcClient.handleError(handle, headers, trailers, err)
+		err = client.grpcClient.handleError(handle, trailers, err)
 		return nil, fmt.Errorf("failed to perform delete, details: %w", err)
 	}
 
@@ -347,13 +347,13 @@ func (client *Client) SubscribeToStream(
 	readClient, err := streamsClient.Read(ctx, subscriptionRequest, callOptions...)
 	if err != nil {
 		defer cancel()
-		err = client.grpcClient.handleError(handle, headers, trailers, err)
+		err = client.grpcClient.handleError(handle, trailers, err)
 		return nil, fmt.Errorf("failed to construct subscription. Reason: %w", err)
 	}
 	readResult, err := readClient.Recv()
 	if err != nil {
 		defer cancel()
-		err = client.grpcClient.handleError(handle, headers, trailers, err)
+		err = client.grpcClient.handleError(handle, trailers, err)
 		return nil, fmt.Errorf("failed to perform read. Reason: %w", err)
 	}
 	switch readResult.Content.(type) {
@@ -401,13 +401,13 @@ func (client *Client) SubscribeToAll(
 	readClient, err := streamsClient.Read(ctx, subscriptionRequest, callOptions...)
 	if err != nil {
 		defer cancel()
-		err = client.grpcClient.handleError(handle, headers, trailers, err)
+		err = client.grpcClient.handleError(handle, trailers, err)
 		return nil, fmt.Errorf("failed to construct subscription. Reason: %w", err)
 	}
 	readResult, err := readClient.Recv()
 	if err != nil {
 		defer cancel()
-		err = client.grpcClient.handleError(handle, headers, trailers, err)
+		err = client.grpcClient.handleError(handle, trailers, err)
 		return nil, fmt.Errorf("failed to perform read. Reason: %w", err)
 	}
 	switch readResult.Content.(type) {

--- a/esdb/impl.go
+++ b/esdb/impl.go
@@ -33,7 +33,14 @@ type grpcClient struct {
 	perRPCCredentials credentials.PerRPCCredentials
 }
 
-func (client *grpcClient) handleError(handle *connectionHandle, headers metadata.MD, trailers metadata.MD, err error) error {
+func (client *grpcClient) handleError(handle *connectionHandle, trailers metadata.MD, err error) error {
+	if client.isClosed() {
+		return &Error{
+			code: ErrorCodeConnectionClosed,
+			err:  fmt.Errorf("connection is closed"),
+		}
+	}
+
 	values := trailers.Get("exception")
 
 	if values != nil && values[0] == "not-leader" {
@@ -55,9 +62,16 @@ func (client *grpcClient) handleError(handle *connectionHandle, headers metadata
 					endpoint:    &endpoint,
 				}
 
-				client.channel <- msg
-				client.logger.error("not leader exception, reconnecting to %v", endpoint)
-				return &Error{code: ErrorCodeNotLeader}
+				if !client.isClosed() {
+					client.channel <- msg
+					client.logger.error("not leader exception, reconnecting to %v", endpoint)
+					return &Error{code: ErrorCodeNotLeader}
+				}
+
+				return &Error{
+					code: ErrorCodeConnectionClosed,
+					err:  fmt.Errorf("connection is closed"),
+				}
 			}
 		}
 	}
@@ -70,19 +84,17 @@ func (client *grpcClient) handleError(handle *connectionHandle, headers metadata
 	client.logger.error("unexpected exception: %v", err)
 
 	code := errToCode(err)
-	if code == ErrorUnavailable {
-		msg := reconnect{
+	if code == ErrorUnavailable && !client.isClosed() {
+		client.channel <- reconnect{
 			correlation: handle.Id(),
 		}
-
-		client.channel <- msg
 	}
 
 	return &Error{code: code, err: err}
 }
 
 func (client *grpcClient) getConnectionHandle() (*connectionHandle, error) {
-	if atomic.LoadInt32(client.closeFlag) != 0 {
+	if client.isClosed() {
 		return nil, &Error{
 			code: ErrorCodeConnectionClosed,
 			err:  fmt.Errorf("connection is closed"),
@@ -102,6 +114,10 @@ func (client *grpcClient) close() {
 		atomic.StoreInt32(client.closeFlag, 1)
 		close(client.channel)
 	})
+}
+
+func (client *grpcClient) isClosed() bool {
+	return atomic.LoadInt32(client.closeFlag) != 0
 }
 
 type getConnection struct {
@@ -254,7 +270,7 @@ func connectionStateMachine(config Configuration, closeFlag *int32, channel chan
 				}
 
 				if state.connection != nil {
-					state.connection.Close()
+					_ = state.connection.Close()
 					state.connection = nil
 				}
 
@@ -589,7 +605,6 @@ func errToCode(err error) ErrorCode {
 }
 
 func shuffleCandidates(src []string) []string {
-	rand.Seed(time.Now().UnixNano())
 	rand.Shuffle(len(src), func(i, j int) {
 		src[i], src[j] = src[j], src[i]
 	})
@@ -598,7 +613,6 @@ func shuffleCandidates(src []string) []string {
 }
 
 func shuffleMembers(src []*gossipApi.MemberInfo) []*gossipApi.MemberInfo {
-	rand.Seed(time.Now().UnixNano())
 	rand.Shuffle(len(src), func(i, j int) {
 		src[i], src[j] = src[j], src[i]
 	})

--- a/esdb/persistent_subscription_client.go
+++ b/esdb/persistent_subscription_client.go
@@ -30,19 +30,19 @@ func (client *persistentClient) ConnectToPersistentSubscription(
 	readClient, err := client.persistentSubscriptionClient.Read(ctx, callOptions...)
 	if err != nil {
 		defer cancel()
-		return nil, client.inner.handleError(handle, headers, trailers, err)
+		return nil, client.inner.handleError(handle, trailers, err)
 	}
 
 	err = readClient.Send(toPersistentReadRequest(bufferSize, groupName, []byte(streamName)))
 	if err != nil {
 		defer cancel()
-		return nil, client.inner.handleError(handle, headers, trailers, err)
+		return nil, client.inner.handleError(handle, trailers, err)
 	}
 
 	readResult, err := readClient.Recv()
 	if err != nil {
 		defer cancel()
-		return nil, client.inner.handleError(handle, headers, trailers, err)
+		return nil, client.inner.handleError(handle, trailers, err)
 	}
 	switch readResult.Content.(type) {
 	case *persistent.ReadResp_SubscriptionConfirmation_:
@@ -77,7 +77,7 @@ func (client *persistentClient) CreateStreamSubscription(
 	defer cancel()
 	_, err := client.persistentSubscriptionClient.Create(ctx, createSubscriptionConfig, callOptions...)
 	if err != nil {
-		return client.inner.handleError(handle, headers, trailers, err)
+		return client.inner.handleError(handle, trailers, err)
 	}
 
 	return nil
@@ -105,7 +105,7 @@ func (client *persistentClient) CreateAllSubscription(
 
 	_, err = client.persistentSubscriptionClient.Create(ctx, protoConfig, callOptions...)
 	if err != nil {
-		return client.inner.handleError(handle, headers, trailers, err)
+		return client.inner.handleError(handle, trailers, err)
 	}
 
 	return nil
@@ -129,7 +129,7 @@ func (client *persistentClient) UpdateStreamSubscription(
 
 	_, err := client.persistentSubscriptionClient.Update(ctx, updateSubscriptionConfig, callOptions...)
 	if err != nil {
-		return client.inner.handleError(handle, headers, trailers, err)
+		return client.inner.handleError(handle, trailers, err)
 	}
 
 	return nil
@@ -153,7 +153,7 @@ func (client *persistentClient) UpdateAllSubscription(
 
 	_, err := client.persistentSubscriptionClient.Update(ctx, updateSubscriptionConfig, callOptions...)
 	if err != nil {
-		return client.inner.handleError(handle, headers, trailers, err)
+		return client.inner.handleError(handle, trailers, err)
 	}
 
 	return nil
@@ -175,7 +175,7 @@ func (client *persistentClient) DeleteStreamSubscription(
 
 	_, err := client.persistentSubscriptionClient.Delete(ctx, deleteSubscriptionOptions, callOptions...)
 	if err != nil {
-		return client.inner.handleError(handle, headers, trailers, err)
+		return client.inner.handleError(handle, trailers, err)
 	}
 
 	return nil
@@ -196,7 +196,7 @@ func (client *persistentClient) DeleteAllSubscription(
 
 	_, err := client.persistentSubscriptionClient.Delete(ctx, deleteSubscriptionOptions, callOptions...)
 	if err != nil {
-		return client.inner.handleError(handle, headers, trailers, err)
+		return client.inner.handleError(handle, trailers, err)
 	}
 
 	return nil
@@ -246,7 +246,7 @@ func (client *persistentClient) listPersistentSubscriptions(
 	resp, err := client.persistentSubscriptionClient.List(ctx, listReq, callOptions...)
 
 	if err != nil {
-		return nil, client.inner.handleError(handle, headers, trailers, err)
+		return nil, client.inner.handleError(handle, trailers, err)
 	}
 
 	var infos []PersistentSubscriptionInfo
@@ -292,7 +292,7 @@ func (client *persistentClient) getPersistentSubscriptionInfo(
 
 	resp, err := client.persistentSubscriptionClient.GetInfo(ctx, getInfoReq, callOptions...)
 	if err != nil {
-		return nil, client.inner.handleError(handle, headers, trailers, err)
+		return nil, client.inner.handleError(handle, trailers, err)
 	}
 
 	info, err := subscriptionInfoFromWire(resp.SubscriptionInfo)
@@ -342,7 +342,7 @@ func (client *persistentClient) replayParkedMessages(
 	_, err := client.persistentSubscriptionClient.ReplayParked(ctx, replayReq, callOptions...)
 
 	if err != nil {
-		return client.inner.handleError(handle, headers, trailers, err)
+		return client.inner.handleError(handle, trailers, err)
 	}
 
 	return nil
@@ -362,7 +362,7 @@ func (client *persistentClient) restartSubsystem(
 	_, err := client.persistentSubscriptionClient.RestartSubsystem(ctx, &shared.Empty{}, callOptions...)
 
 	if err != nil {
-		return client.inner.handleError(handle, headers, trailers, err)
+		return client.inner.handleError(handle, trailers, err)
 	}
 
 	return nil

--- a/esdb/projection_client.go
+++ b/esdb/projection_client.go
@@ -433,7 +433,7 @@ func (client *ProjectionClient) listInternal(
 
 		if err != nil {
 			if !errors.Is(err, io.EOF) {
-				err = client.inner.grpcClient.handleError(handle, headers, trailers, err)
+				err = client.inner.grpcClient.handleError(handle, trailers, err)
 				return nil, err
 			}
 

--- a/esdb/reads.go
+++ b/esdb/reads.go
@@ -48,7 +48,7 @@ func (stream *ReadStream) Recv() (*ResolvedEvent, error) {
 		atomic.StoreInt32(stream.closed, 1)
 
 		if !errors.Is(err, io.EOF) {
-			err = stream.params.client.handleError(stream.params.handle, *stream.params.headers, *stream.params.trailers, err)
+			err = stream.params.client.handleError(stream.params.handle, *stream.params.trailers, err)
 		}
 
 		return nil, err

